### ISSUE TITLE
Implement auto fix functionality for semi-spacing rule

### ISSUE THREAD

**Issue by alberto** [eslint/eslint#3829]
Implement auto fix functionality for semi-spacing rule

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
I'd like to hold off on this one for a little bit before implementing. I'm a bit concerned about overlap with the `semi` rule fixes. So I'll mark as accepted, but would like to wait a release or two before merging.

**Comment by alberto** [comment]
@nzakas are you ok with me implementing any other spacing fixes or do you want to put all of them on hold for now?

**Comment by nzakas** [comment]
Just this one. However, if you want to implement auto-fixing in a rule, file an issue and wait for it to be accepted before starting work. We are intentionally going slow on this to account for bugs we will find a long the way, so we don't want to add a lot of auto-fixing rules at once.

**Comment by alberto** [comment]
Understood

**Comment by alberto** [comment]
@nzakas what's your take on this? I don't think the fix introduces any overlap.

**Comment by nzakas** [comment]
Can you explain why you think there's no overlap?

**Comment by nzakas** [comment]
Or rather, no conflicts?

**Comment by alberto** [comment]
I think there is already a possible conflict, where a semi is needed and autofixed by the `semi` rule, but not according to the config of `semi-spacing`. Adding autofixing for `semi-spacing` doesn't fix this, but I don't see how it would create any other conflict. Maybe you see something that I am missing here.

**Comment by nzakas** [comment]
We shouldn't add more conflicts just because some already exist. What I'm looking for here is examples showing this would not conflict or that it would.

**Comment by alberto** [comment]
The thing is I don't think adding the fix causes a conflict, but giving a couple of examples showing no conflicts doesn't prove that. When do you think they would cause a conflict?

**Comment by alberto** [comment]
The only case I can think of is if you set `semi` to `never` and you also have `semi-spacing` enabled. It could cause conflicts if you have unneeded semicolons. But it doesn't make sense to me to enable `semi-spacing` in that case, since it wouldn't do anything.

**Comment by nzakas** [comment]
Theoretically, any inserted semicolon could also require a space before it. Those warnings would be missed when the semicolon is inserted. I suppose we could limit the fix to be only when there is no space required before the semicolon?

**Comment by alberto** [comment]
Yes it could, but that is already happening (it's the one I was refering to [here](https://github.com/eslint/eslint/issues/3829#issuecomment-164908473)). That's caused by `semi`'s autofix, not by the addition of the fix to `semi-spacing`.

**Comment by nzakas** [comment]
You've lost me, what do you mean "that is already happening"? What is already happening?

**Comment by alberto** [comment]
What you just said. :) `semi` is already inserting missing semicolons. If `semi-spacing` is configured to require a space before the semicolon, that warning will be missed.

**Comment by nzakas** [comment]
Ah, good point. Okay, go for it.
